pkg/store/tsdb.go

```diff
@@ -63,21 +73,68 @@ type ReadWriteTSDBStore struct {
 
 // NewTSDBStore creates a new TSDBStore.
 // NOTE: Given lset has to be sorted.
-func NewTSDBStore(logger log.Logger, db TSDBReader, component component.StoreAPI, extLset labels.Labels) *TSDBStore {
+func NewTSDBStore(logger log.Logger, db TSDBReader, component component.StoreAPI, extLset labels.Labels, metricNameFilterEnabled bool) *TSDBStore {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
-	return &TSDBStore{
+
+	var (
+		storeFilter       filter.StoreFilter
+		startFilterUpdate bool
+	)
+
+	storeFilter = filter.AllowAllStoreFilter{}
+	if metricNameFilterEnabled {
+		startFilterUpdate = true
+		storeFilter = filter.NewCuckooMetricNameStoreFilter(1000000) // about 1MB on 64bit machines.
+	}
+
+	st := &TSDBStore{
 		logger:           logger,
 		db:               db,
 		component:        component,
 		extLset:          extLset,
+		storeFilter:      storeFilter,
 		maxBytesPerFrame: RemoteReadFrameLimit,
+		close:            func() {},
 		buffers: sync.Pool{New: func() interface{} {
 			b := make([]byte, 0, initialBufSize)
 			return &b
 		}},
 	}
+
+	if startFilterUpdate {
+		t := time.NewTicker(15 * time.Second)
+		ctx, cancel := context.WithCancel(context.Background())
+		updateFilter := func() {
+			vals, err := st.LabelValues(context.Background(), &storepb.LabelValuesRequest{
```

Comment:
Should we use the ctx variable here so that we can cancel the request on shutdown?

I would do something like this and pass the `ctx` when calling `updateFilter`:
```suggestion
		updateFilter := func(ctx context.Context) {
			vals, err := st.LabelValues(ctx, &storepb.LabelValuesRequest{
```

Reply:
done, lets pass context